File: bam_backup/js/bam_backup_page.js
// Copyright 2021 BlueCat Networks (USA) Inc. and its affiliates. All Rights Reserved.
$(document).ready(function () {
    $('#deleteSelected').prop('disabled', true)
    getBackupData();
});


function getBackupData(isSkipCheck) {
    $.ajax({
        url: `/api/v1/bam_backup/backup_file`,
        type: 'get',
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            var bkDataHTML = [];
            if (data.length == 0 & !isSkipCheck) {
                $('#statusMsg').addClass('inform');
                $("#statusMsg").text('BAM Backup Data not found.');
                return false
            }

            for (var i = 0; i < data.length; i++) {
                bkDataHTML.push(`<tr><td>
                                    <input type="checkbox" id="ck_${data[i]['name']}" class="input-track" onclick="checkDeleteSelected()">
                                    <div style="margin-top: 5px;">${data[i]['name']}</div>
                                </td>
                                <td>${data[i]['date']}</td>
                                <td>${data[i]['size']}</td>
                                <td>
                                    <i class="fa fa-download" aria-hidden="true" id='downBackup_${data[i].name}' onclick='downBackupFile("${data[i].name}")'></i>
                                </td></tr>`)
            }
            $('#bkDataTable > tbody').html(bkDataHTML);
            $('#runBackup').prop("disabled", false);
            if (data.length >= MAX_COUNT_FILE) {
                definedErrorMessage(`The number of backup files have reached the maximum file count (${MAX_COUNT_FILE}). Please delete to continue run backup.`)
                $('#runBackup').prop("disabled", true);
            }
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}


function downBackupFile(bkName) {
    var $content = DEFAULT_CONTENT;
    $('body').prepend($content);
    var href = window.location.origin + '/api/v1/bam_backup/local_backup_path/' + bkName
    window.location.href = href;
    setTimeout(
        function () {
            $("#wait").remove();
        }, 3500);

}


function checkDeleteSelected() {
    ($("input[type=checkbox][id^='ck_']:checked").length > 0) ? $('#deleteSelected').prop('disabled', false): $('#deleteSelected').prop('disabled', true);
}


function runBackup() {
    $.ajax({
        url: `/api/v1/bam_backup/run`,
        type: 'post',
        dataType: 'json',
        contentType: 'application/json',
        success: function (result) {
            getBackupData(true);
            (result.includes('successfully') ? definedSuccessMessage(result) : definedErrorMessage(result));
        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}


function checkBackupStatus() {
    $.ajax({
        url: `/api/v1/bam_backup/check_status`,
        type: 'get',
        dataType: 'json',
        contentType: 'application/json',
        success: function (data) {
            (data.length > 0) ? $('#checkBKStatus').show(): $('#checkBKStatus').hide();
            var bkStatusHTML = [];
            for (var i = 0; i < data.length; i++) {
                bkStatusHTML.push(`<tr>
                                    <td>${data[i]['name']}</td>
                                    <td>${data[i]['start_time']}</td>
                                    <td>${data[i]['end_time']}</td>
                                    <td>${data[i]['status']}</td>
                                </tr>`)
            }
            $('#bkStatusTable > tbody').html(bkStatusHTML);
            definedSuccessMessage('Check backup status successfully.');

        },
        error: function (message) {
            definedErrorMessage(message)
        }
    });
}



function deleteBackupFile(bkName) {
    var bkList = [bkName];
    if (!bkName) {
        var bkList = [];
        $("input[type=checkbox][id^='ck_']:checked").each(function (element, index) {
            var bkName = $(this).attr("id").replace('ck_', '');
            bkList.push(bkName);
        })
    }

    var bkListHTML = '';
    for (var j =0;j<bkList.length;j++){
        bkListHTML += `<li style='margin-left:30px'>${bkList[j]}</li>`
    }

    var $content = `<div class='dialog-ovelay'>
                        <div class='dialog'>
                            <header>
                                <h3>Confirm Delete Backup File(s)</h3><i class='fa fa-close'></i>
                            </header>
                            <div class='dialog-msg' style='margin-bottom:5px;'>
                                <p>Backup File(s): <br> 
                                <ul style="list-style-type:square">${bkListHTML}</ul> will be deleted!</p>
                                <p>Do you still want to continue ?</p>
                            </div>
                            <footer>
                                <div class='controls' style='margin-left:5px'>
                                    <button class='button button-default btn-primary doAction' style='margin-right:10px!important'>Yes</button>
                                    <button class='button button-danger btn-default cancelAction' style='float: unset!important;'>No</button>
                                </div>
                            </footer>
                        </div>
                    </div>`;
    $('body').prepend($content);
    $('.doAction').click(function () {
        $(this).parents('.dialog-ovelay').fadeOut(500, function () {
            $(this).remove();
        });

        $.ajax({
            url: `/api/v1/bam_backup/backup_file`,
            type: 'delete',
            dataType: 'json',
            data: JSON.stringify({
                'backup_files_name': bkList
            }),
            contentType: 'application/json',
            success: function (result) {
                getBackupData(true);
                (result.includes('successfully') ? definedSuccessMessage(result) : definedErrorMessage(result));
            },
            error: function (message) {
                definedErrorMessage(message)
            }
        });

    });
    $('.cancelAction, .fa-close').click(function () {
        $(this).parents('.dialog-ovelay').fadeOut(500, function () {
            $(this).remove();
        });
    });


}

File: bam_backup/bam_backup_page.py
# Copyright 2020 BlueCat Networks (USA) Inc. and its affiliates. All Rights Reserved.
import os
import sys
import traceback
from flask_restplus import Resource
from flask import render_template, g, make_response, jsonify, send_file

from bluecat import route, util
import config.default_config as config
from main_app import app, api
from ..common.exceptions import UserException
from ..common.backup_file import RemoteBAMBackupClient
from ..common.common import get_config
from ..common.constants import DEFAULT_BAM_BACKUP_CONFIG
from .libs.models import bam_backup_models
from .libs.parsers import bam_backup_parsers


def module_path():
    encoding = sys.getfilesystemencoding()
    return os.path.dirname(os.path.abspath(__file__))


@route(app, '/bam_backup/bam_backup_endpoint')
@util.workflow_permission_required('bam_backup_page')
@util.exception_catcher
@util.ui_secure_endpoint
@util.no_cache
def bam_backup_bam_backup_page():
    config_file = get_config(DEFAULT_BAM_BACKUP_CONFIG)
    return render_template(
        'bam_backup_page.html',
        text=util.get_text(module_path(), config.language),
        options=g.user.get_options(),
        max_count_file=config_file.get("BAM_BACKUP", "max_count_file", fallback="")
    )


bam_backup_ns = api.namespace(
    'bam_backup',
    path='/bam_backup',
    description='Bam Backup by Gateway',
)


@bam_backup_ns.route('/backup_file')
class BAMBackupFiles(Resource):
    @util.workflow_permission_required('bam_backup_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self):
        """ Get backup file from BAM"""
        try:
            remote_client = RemoteBAMBackupClient(g.user.get_api_netloc())
            result = remote_client.read_backup_file()
            remote_client.disconnect()
            return jsonify(result)
        except UserException as user_ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(user_ex)), 400)
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)

    @util.workflow_permission_required('bam_backup_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    @api.expect(bam_backup_models.delete_bam_backup_model, validate=True)
    def delete(self):
        """ Delete backup file from BAM"""
        try:
            data = bam_backup_parsers.delete_bam_backup_parser.parse_args()
            backup_files_name = data.get('backup_files_name')
            remote_client = RemoteBAMBackupClient(g.user.get_api_netloc())
            result = remote_client.delete_backup_file(backup_files_name)
            return jsonify(result)
        except UserException as user_ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(user_ex)), 400)
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@bam_backup_ns.route('/backup_file/<string:backup_file_name>')
class BAMBackup(Resource):
    @util.workflow_permission_required('bam_backup_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self, backup_file_name):
        """ Support get backup file from BAM by backup name"""
        try:
            hostname = g.user.get_api_netloc()
            remote_client = RemoteBAMBackupClient(hostname)
            bk_list = remote_client.read_backup_file()
            filter_bk = [bk for bk in bk_list if bk.get('name') == backup_file_name]
            result = filter_bk[0] if filter_bk else {}
            return jsonify(result)
        except UserException as user_ex:
            remote_client = None
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(user_ex)), 400)
        except Exception as e:
            remote_client = None
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)
        finally:
            if remote_client:
                remote_client.disconnect()


@bam_backup_ns.route('/local_backup_path/<string:backup_file_name>')
class DownloadBAMBackup(Resource):
    @util.workflow_permission_required('bam_backup_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self, backup_file_name):
        """ Support download backup file from BAM and return local path"""
        try:
            hostname = g.user.get_api_netloc()
            remote_client = RemoteBAMBackupClient(hostname)
            local_bk_path = remote_client.download_and_get_local_backup_file(backup_file_name)
            return send_file(local_bk_path, as_attachment=True), 200
        except UserException as user_ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(user_ex)), 400)
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@bam_backup_ns.route('/run')
class RunBAMBackup(Resource):
    @util.workflow_permission_required('bam_backup_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def post(self):
        """ Run default backup file"""
        try:
            remote_client = RemoteBAMBackupClient(g.user.get_api_netloc())
            result = remote_client.run_backup_file()
            return jsonify(result)
        except UserException as user_ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(user_ex)), 400)
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)


@bam_backup_ns.route('/check_status')
class CheckBackupStatus(Resource):
    @util.workflow_permission_required('bam_backup_page')
    @util.exception_catcher
    @util.ui_secure_endpoint
    @util.no_cache
    def get(self):
        """ Check status after run backup file"""
        try:
            remote_client = RemoteBAMBackupClient(g.user.get_api_netloc())
            result = remote_client.check_backup_status()
            return jsonify(result)
        except UserException as user_ex:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(user_ex)), 400)
        except Exception as e:
            g.user.logger.error(traceback.format_exc())
            return make_response(jsonify(str(e)), 500)
